catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,605 @@
-  private ScheduledExecutorService scheduler;
+  private List<FederatedSource> federatedSources = Collections.emptyList();
 
-  private ScheduledFuture<?> handle;
+  private List<CatalogProvider> catalogProviders = Collections.emptyList();
 
-  private SourcePollerRunner runner;
+  private List<CatalogStore> catalogStores = Collections.emptyList();
+
+  private final ExecutorService sourceAvailabilityChecksThreadPool;
+
+  private final Cache<SourceKey, CachedSourceAvailability> cache;
+
+  @Nullable private ExecutorService pollingExecutorService;
+
+  /** Does not start polling */
+  public SourcePoller() {
+    final ThreadFactory threadFactory =
+        StandardThreadFactoryBuilder.newThreadFactory("sourcePollerScheduledPollingThread");
+    sourceAvailabilityChecksThreadPool = Executors.newCachedThreadPool(threadFactory);
+
+    cache =
+        CacheBuilder.newBuilder()
+            .maximumSize(1000)
+            // TODO Expire orphaned entries in the map caused by when sources are modified and their
+            // key
+            // is different
+            .removalListener(
+                removalNotification ->
+                    LOGGER.debug(
+                        "Removing SourceKey {} from cache. cause={}",
+                        removalNotification.getKey(),
+                        removalNotification.getCause()))
+            .build();
+  }
+
+  public void destroy() {
+    if (pollingExecutorService != null) {
+      MoreExecutors.shutdownAndAwaitTermination(
+          pollingExecutorService,
+          EXECUTOR_SERVICE_TERMINATION_TIMEOUT,
+          EXECUTOR_SERVICE_TERMINATION_TIMEOUT_TIME_UNIT);
+    }
+
+    MoreExecutors.shutdownAndAwaitTermination(
+        sourceAvailabilityChecksThreadPool,
+        EXECUTOR_SERVICE_TERMINATION_TIMEOUT,
+        EXECUTOR_SERVICE_TERMINATION_TIMEOUT_TIME_UNIT);
+  }
 
   /**
-   * Constructor to schedule the SourcePollerRunner to execute immediately and at a fixed interval,
-   * currently set at every 1 minute. This constructor is invoked by the CatalogFramework's
-   * blueprint.
+   * This method is an non-blocking alternative to {@link Source#isAvailable()} which provides more
+   * more details about the availability than just available or unavailable. The last availability
+   * check is guaranteed to be started when the source is bound with {@link #bind(Source)} or at the
+   * start of the lat poll interval, configured by {@link #setPollIntervalMinutes(long)}.
    *
-   * @param incomingRunner the SourcePollerRunner to use for polling
+   * @return a {@link SourceAvailability} for the availability of the {@code source}, or {@link
+   *     Optional#empty()} if the availability of the {@code source} has not yet been checked
+   * @throws IllegalArgumentException if the {@code source} is {@code null}
    */
-  public SourcePoller(SourcePollerRunner incomingRunner) {
+  public Optional<SourceAvailability> getSourceAvailability(final Source source) {
+    notNull(source);
```

Comment:
Doesn't this throw a NullPointerException? The javadoc above makes it sound like it will throw an IllegalArgumentException. Looks like all the places it is used in this pr have the same javadoc issue.

Reply:
N/A after new changes